go-controller/pkg/ovn/obj_retry.go

```diff
@@ -1515,10 +1515,19 @@ func (oc *Controller) WatchResource(objectsToRetry *retryObjs) (*factory.Handler
 				}
 
 				// skip the whole update if the new object doesn't exist anymore in the API server
-				newer, err = oc.getResourceFromInformerCache(objectsToRetry.oType, newKey)
+				latest, err := oc.getResourceFromInformerCache(objectsToRetry.oType, newKey)
 				if err != nil {
-					klog.Warningf("Unable to get %s %s from informer cache (perhaps it was already"+
-						" deleted?), skipping update: %v", objectsToRetry.oType, newKey, err)
+					// When processing an object in terminal state there is a chance that it was already removed from
+					//  the API server. Since delete events for objects in terminal state are skipped delete it here.
+					// This only applies to pod watchers (pods + dynamic network policy handlers watching pods).
+					if kerrors.IsNotFound(err) && oc.isObjectInTerminalState(objectsToRetry.oType, newer) {
+						klog.Warningf("%s %s is in terminal state but no longer exists in informer cache, removing",
+							objectsToRetry.oType, newKey)
+						oc.processObjectInTerminalState(objectsToRetry, newer, newKey, resourceEventUpdate)
+					} else {
+						klog.Warningf("Unable to get %s %s from informer cache (perhaps it was already"+
+							" deleted?), skipping update: %v", objectsToRetry.oType, newKey, err)
+					}
```

Comment:
This is exactly what we were looking for, but by running the whole update logic on `latest`, we now invalidate the output of `areResourcesEqual`, which is useful for network policies, services referred to by network policies, nodes, and egress firewalls.

I would go as far as overwriting `newer` with `latest` as soon as we've looked up the pod status in newer and move it to the top of `UpdateFunc`:

https://github.com/ricky-rav/ovn-kubernetes/pull/8/commits/b208954b25a349664d95763f840b5eea1b0077de

Reply:
After a short discussion with @tssurya I wanted to leave the comparison as is for now to avoid any side effects.

If you guys think it is ok to compare `old` with the latest object from the informer cache I can make that change.
@trozet what do you think?

Reply:
This is a bit tricky :) So my instinct is to say lets be level driven and take the latest version to do everything with. However, the update method itself and how we use it can be event based. For example, rather than just saying is old different from new, lets change the state to be new...sometimes (like in network policy) we say ok lets delete old and configure new.

That being said lets take a case like a pod or some resource that is updated several times like a pod being scheduled or annotated. Lets call it object A and say its initial create state is 0, and its updated 3 times, and now we are processing it after the updates. Let's compare the level driven way (comparing latest to old):
1. Event old 0, new 1. latest is new 3. Delete old 0, configure new 3.
2. Event old 1, new 2. latest is new 3. Delete old 1, configure new 3.
3. Event old 2, new 3. latest is new 3. Delete old 2, configure new 3.

You can see that even with updating to latest, we still do extra pointless config in this case. If you consider a case where new 1 and new 2 would not contain changes that would trigger an update then doing the above method would actually cost more than doing an event driven approach:

1. Event old 0, new 1. latest is new 3, but compare 0 to 1, no change.
2. Event old 1, new 2. latest is new 3, but compare 1 to 2, no change.
3. Event old 2, new 3. compare 2 to 3, change. Delete old 2, configure new 3.

Since we have no way to compare last configured thing with latest thing trying to be level driven here doesn't seem to yield any real benefit to me. One could argue that if we compared what was configured in OVN with the latest to be configured then that would be more efficient as less operations to OVN, and maybe we do that in some places (like a mutate operation to include an address in a set, or update a route). However, I think the majority of cases to determine if there is an update is going to be comparing the old version in each event.

So this is a long winded way of saying I think we should only configure the new version we are given in the event. I'm open to other opinions though.